File: trabajos-elixir/trabajos-java-recursividad/src/PalindromoRecursivo.java
public class PalindromoRecursivo {
    public static void main(String[] args) {
        System.out.println("Detectar palindromo");
        boolean resultado = esPalindromo("aceite");
        System.out.println( resultado? "Es un palindromo" : "No es un palindromo");
    }

    public static boolean esPalindromo(String palabra) {
        return esPalindromoAux(palabra, 0, palabra.length());
    }

    public static boolean esPalindromoAux(String palabra, int inicio, int fin) {
        // Caso base (Condicion de salida)
        if (inicio >= fin) {
            return true;
        } else if (palabra.charAt(inicio) != (palabra.charAt(fin-1))) {
            return false;
        }
        
        // Caso recursivo
        return esPalindromoAux(palabra, inicio+1, fin-1);
    }
}

File: trabajos-elixir/trabajos-elixir-recursividad/palindromo_recursivo.exs
# Palindromo
defmodule PalindromoRecursivo do
  # si regresa booleano acaba con ?
  def es_palindromo?(palabra) when is_binary(palabra) do
    es_palindromo_aux(palabra)
  end

  # Caso base (Condiciones de salida)
  defp es_palindromo_aux(""), do: true
  defp es_palindromo_aux(<<_un_caracter::utf8>>), do: true

  # Caso recursivo
  defp es_palindromo_aux(palabra) do
    primero = String.at(palabra, 0)
    ultimo = String.at(palabra, -1)

    if primero == ultimo do
      subcadena = String.slice(palabra, 1..-2//1)
      es_palindromo_aux(subcadena)
    else
      false
    end
  end
end

mensaje = if PalindromoRecursivo.es_palindromo?("rallar") do
   "Es un palindromo"
else
   "No es un palindromo"
end

IO.puts(mensaje)
